mono/profiler/mono-profiler-log.c

```diff
@@ -652,29 +729,31 @@ buffer_unlock (void)
 {
 	mono_memory_barrier ();
 
+	gint32 state = InterlockedRead (&buffer_lock_state);
+
 	// See the comment in buffer_lock ().
-	if (InterlockedReadPointer (&buffer_rwlock_exclusive) == (gpointer) thread_id ())
+	if (state == PROF_TLS_GET ()->small_id << 16)
 		return;
 
-	g_assert (InterlockedRead (&buffer_rwlock_count) && "Why are we trying to decrement a zero reader count?");
+	g_assert (state && "Why are we decrementing a zero reader count?");
+	g_assert (!(state >> 16) && "Why is the exclusive lock held?");
 
-	InterlockedDecrement (&buffer_rwlock_count);
+	InterlockedDecrement (&buffer_lock_state);
 }
 
 // Cannot be used recursively.
 static void
 buffer_lock_excl (void)
 {
-	gpointer tid = (gpointer) thread_id ();
+	gint32 new_ = PROF_TLS_GET ()->small_id << 16;
 
-	g_assert (InterlockedReadPointer (&buffer_rwlock_exclusive) != tid && "Why are we taking the exclusive lock twice?");
+	g_assert (InterlockedRead (&buffer_lock_state) != new_ && "Why are we taking the exclusive lock twice?");
 
-	MONO_ENTER_GC_SAFE;
+	InterlockedWrite (&buffer_lock_exclusive_intent, 1);
```

Comment:
This write bias won't work reliably across multiple writers as buffer_unlock_excl zeroes buffer_lock_exclusive_intent.

It might make sense as a way to provide some chance to readers in case of too many writers, but it won't make the biasing as good as you'd expect. What do you think is the best option here?